### DISCUSSION TITLE
Subscriptions and interfaces

### DISCUSSION THREAD

**Discussion by theogravity** [hayes/pothos#174] [Q&A] [answered]
I am currently porting over a subscription from our apollo server to giraph. We have a single resolver for our subscription, which can emit different kinds of events, which all implement an interface.

It looks something like this:

```ts
export class BaseSampleEvent {
  id: string;

  constructor(id: string) {
    this.id = id;
  }
}

builder.interfaceType(BaseSampleEvent, {
  name: 'IBaseSampleEvent',
  fields: (t) => ({
    id: t.exposeID('id'),
  }),
});

export class NewUserEvent extends BaseSampleEvent {
   ...
}

builder.objectType(NewUserEvent, {
  name: 'NewUserEvent',
  interfaces: [BaseTeamEvent],
  isTypeOf: (value) => value instanceof NewUserEvent,
  fields: (t) => ({
    id: t.exposeID('id'),
  }),
});

builder.subscriptionField('sampleEvents', (t) => {
  return t.field({
    type: BaseSampleEvent
    subscribe: () => {
       // from graphql-subscriptions lib
       return withFilter(...)()
    },
    resolve: (payload) => payload as BaseSampleEvent
  })
})
```

When I attempt to publish the `NewUserEvent` event, I get the following error:

```
ERROR 20:17:8.51 Abstract type "IBaseSampleEvent" must resolve to an Object type at runtime for field "Subscription.sampleEvents" with value { ... } Either the "sampleEvents" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.\n    at ensureValidRuntimeType 
```

Not sure what I'm doing wrong here as I do have `isTypeOf` defined. In our apollo implementation, we did have to define a set of mapped resolvers / types for `BaseSampleEvent`, and the closest thing I'm seeing with this is to use a `giraphql` union, and assign that union as the return type of the subscription.

If that is the case, do I need the `isTypeOf` on the `builder.objectType(NewUserEvent)` event objects themselves?

**Comment by hayes** [comment]
Are your events actually instances of the classes you are defining, or are they just objects with the same properties?

**Reply by hayes** [reply]
I think you would see this error if all your isTypeOf functions are returning false

**Reply by hayes** [reply]
I think isTypeOf can actually be optional too because the default is an instanceof check when you use a class as the type ref

**Comment by theogravity** [comment, marked as answer]
They're actually instances of the classes. It seems I can't remove `isTypeOf` without removing the interfaces property. If I remove both, the event types do not implement anything in the generated schema.

But, I think I figured out what's going on. I noticed this behavior too when it came to writing the subscribe function in our apollo server.

The `value` passed to `isTypeOf` is serialized to an object. When I console.logged out the event in `isTypeOf`, it was a pure object. I wonder if the serialization is a native gql thing or specific to the pubsub lib being used.

We actually add in an `eventType` property in our events, and so when I compared that property, it works!

```
  isTypeOf: (value) => {
    return value.eventType === EventType.NewUserEvent;
  },
```

**Reply by hayes** [reply]
I think it must be a pub-sub thing.  I know gql won't automatically serialize it.  Probably to standardize behavior across different implementations (in prod most people will use something like redis, which would cause there data to be serialized, so serialization would make things consistent for an in-process implementation)
